# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1707947459 -3600
# Node ID 76ace2e9148c46dba9f35ddcf1a8b336e41948e5
# Parent  ee3b7b6e0a00b1b4484e814e771b4952411c7216
Bug 531319 - Fix opening a message in conversation from a standalone window. r=vineet

This opens a new 3pane window if necessary, when a message displayed in a standalone window is
to be shown in a conversation.

Differential Revision: https://phabricator.services.mozilla.com/D200063

diff --git a/mail/modules/ConversationOpener.jsm b/mail/modules/ConversationOpener.jsm
--- a/mail/modules/ConversationOpener.jsm
+++ b/mail/modules/ConversationOpener.jsm
@@ -41,29 +41,53 @@ class ConversationOpener {
   onItemsModified(items) {}
   onItemsRemoved(items) {}
   onQueryCompleted(collection) {
     try {
       if (!collection.items.length) {
         console.error("Couldn't find a collection for msg: " + this._msgHdr);
       } else {
         let message = collection.items[0];
-        let tabmail = this.window.top.document.getElementById("tabmail");
-        if (!tabmail) {
-          tabmail = Services.wm
-            .getMostRecentWindow("mail:3pane")
-            .document.getElementById("tabmail");
-        }
-        tabmail.openTab("mail3PaneTab", {
+        const tabType = "mail3PaneTab";
+        const tabParams = {
           folderPaneVisible: false,
           syntheticView: new GlodaSyntheticView({
             conversation: message.conversation,
             message,
           }),
           title: message.conversation.subject,
           background: false,
-        });
+        };
+        const tabmail =
+          this.window.top.document.getElementById("tabmail") ??
+          Services.wm
+            .getMostRecentWindow("mail:3pane")
+            ?.document.getElementById("tabmail");
+        if (tabmail) {
+          tabmail.openTab(tabType, tabParams);
+        } else {
+          const win = Services.ww.openWindow(
+            null,
+            "chrome://messenger/content/messenger.xhtml",
+            "_blank",
+            "chrome,dialog=no,all",
+            Cc["@mozilla.org/supports-string;1"].createInstance(
+              Ci.nsISupportsString
+            )
+          );
+          const mailStartupObserver = {
+            observe(subject) {
+              if (subject == win) {
+                win.document
+                  .getElementById("tabmail")
+                  .openTab(tabType, tabParams);
+                Services.obs.removeObserver(this, "mail-startup-done");
+              }
+            },
+          };
+          Services.obs.addObserver(mailStartupObserver, "mail-startup-done");
+        }
       }
     } catch (e) {
       console.error(e);
     }
   }
 }
